File: ui/profile_select.py
"""Launch screen: pick an existing profile or create a new one."""

import tkinter as tk

from morse import storage
from ui import theme


class ProfileSelect(tk.Frame):
    def __init__(self, parent, app):
        super().__init__(parent, bg=theme.BG)
        self.app = app
        self._build()

    def _build(self):
        tk.Label(
            self, text="DitDash", font=theme.TITLE_FONT, bg=theme.BG, fg=theme.ACCENT
        ).pack(pady=(40, 4))
        tk.Label(
            self,
            text="Morse code trainer",
            font=theme.SMALL,
            bg=theme.BG,
            fg=theme.MUTED,
        ).pack(pady=(0, 24))

        tk.Label(
            self, text="Choose a profile", font=theme.HEADING, bg=theme.BG, fg=theme.FG
        ).pack(pady=(0, 8))

        list_frame = tk.Frame(self, bg=theme.BG)
        list_frame.pack(fill="x", padx=48)

        profiles = storage.list_profiles()
        if profiles:
            for name in profiles:
                theme.button(
                    list_frame,
                    name,
                    lambda n=name: self._choose(n),
                    anchor="w",
                ).pack(fill="x", pady=4)
        else:
            tk.Label(
                list_frame,
                text="No profiles yet — create one below.",
                font=theme.SMALL,
                bg=theme.BG,
                fg=theme.MUTED,
            ).pack(pady=8)

        tk.Frame(self, bg=theme.BTN_BG, height=1).pack(fill="x", padx=48, pady=24)

        tk.Label(
            self, text="New profile", font=theme.HEADING, bg=theme.BG, fg=theme.FG
        ).pack()
        entry_row = tk.Frame(self, bg=theme.BG)
        entry_row.pack(pady=8)
        self.name_var = tk.StringVar()
        entry = tk.Entry(
            entry_row,
            textvariable=self.name_var,
            font=theme.FONT,
            width=18,
            bg=theme.PANEL,
            fg=theme.FG,
            insertbackground=theme.FG,
            relief="flat",
        )
        entry.pack(side="left", ipady=6, padx=(0, 8))
        entry.bind("<Return>", lambda _e: self._create())
        theme.button(entry_row, "Create", self._create, bg=theme.ACCENT, fg=theme.BG).pack(
            side="left"
        )

        self.error = tk.Label(self, text="", font=theme.SMALL, bg=theme.BG, fg=theme.BAD)
        self.error.pack(pady=6)

    def _choose(self, name):
        self.app.load_profile(name)
        from ui.main_menu import MainMenu

        self.app.show(MainMenu)

    def _create(self):
        name = self.name_var.get().strip()
        if not name:
            self.error.config(text="Please enter a name.")
            return
        if name in storage.list_profiles():
            self.error.config(text="That profile already exists.")
            return
        storage.create_profile(name)
        self._choose(name)

File: web/js/profileSelect.js
// Launch screen: pick an existing profile or create a new one.
//
// Profiles may optionally carry a light PIN — not real security, just a
// soft deterrent so one person's profile isn't casually poked at by
// another person using the same device.

import * as storage from "./storage.js";
import { el, button } from "./dom.js";

export class ProfileSelect {
  constructor(root, app) {
    this.root = root;
    this.app = app;
    this._build();
  }

  _build() {
    const wrap = el("div", { class: "screen center" });

    wrap.appendChild(el("div", { class: "title", text: "DitDash" }));
    wrap.appendChild(el("p", { class: "small muted", text: "Morse code trainer" }));
    wrap.appendChild(el("h2", { class: "heading", text: "Choose a profile" }));

    const list = el("div", { class: "profile-list" });
    const profiles = storage.listProfiles();
    if (profiles.length) {
      for (const name of profiles) {
        const locked = !!storage.loadProfile(name).pin;
        const label = locked ? `${name}  🔒` : name;
        list.appendChild(button(label, () => this._choose(name), "btn-block btn-left"));
      }
    } else {
      list.appendChild(el("p", { class: "small muted", text: "No profiles yet — create one below." }));
    }
    wrap.appendChild(list);

    this.pinPrompt = el("div", { class: "pin-prompt", style: { display: "none" } });
    wrap.appendChild(this.pinPrompt);

    wrap.appendChild(el("div", { class: "divider" }));

    wrap.appendChild(el("h2", { class: "heading", text: "New profile" }));
    const entryRow = el("div", { class: "entry-row" });
    const input = el("input", {
      class: "text-input",
      type: "text",
      placeholder: "Name",
      "aria-label": "New profile name",
    });
    const pinInput = el("input", {
      class: "text-input pin-field",
      type: "password",
      inputmode: "numeric",
      maxlength: "8",
      placeholder: "PIN (optional)",
      "aria-label": "New profile PIN, optional",
    });
    input.addEventListener("keydown", (e) => {
      if (e.key === "Enter") this._create();
    });
    pinInput.addEventListener("keydown", (e) => {
      if (e.key === "Enter") this._create();
    });
    entryRow.appendChild(input);
    entryRow.appendChild(pinInput);
    entryRow.appendChild(button("Create", () => this._create(), "btn-accent"));
    wrap.appendChild(entryRow);

    this.error = el("p", { class: "small error", "aria-live": "polite" });
    wrap.appendChild(this.error);

    this.root.appendChild(wrap);
    this.input = input;
    this.pinInput = pinInput;
  }

  _choose(name) {
    const profile = storage.loadProfile(name);
    if (profile.pin) {
      this._showPinPrompt(name, profile);
    } else {
      this._enter(name);
    }
  }

  _showPinPrompt(name, profile) {
    this.pinPrompt.innerHTML = "";
    this.pinPrompt.style.display = "flex";

    this.pinPrompt.appendChild(el("span", { class: "small", text: `PIN for ${name}` }));
    const input = el("input", {
      class: "text-input pin-field",
      type: "password",
      inputmode: "numeric",
      maxlength: "8",
    });
    const errorLbl = el("span", { class: "small error", "aria-live": "polite" });

    const submit = () => {
      if (storage.verifyPin(profile, input.value)) {
        this.pinPrompt.style.display = "none";
        this._enter(name);
      } else {
        errorLbl.textContent = "Wrong PIN.";
        input.value = "";
        input.focus();
      }
    };
    input.addEventListener("keydown", (e) => {
      if (e.key === "Enter") submit();
    });

    this.pinPrompt.appendChild(input);
    this.pinPrompt.appendChild(button("Unlock", submit, "btn-accent"));
    this.pinPrompt.appendChild(errorLbl);
    input.focus();
  }

  _enter(name) {
    // Goes through app.loadProfile() rather than assigning profileName/profile
    // directly, so it also syncs the "Keep headphones awake" keep-alive tone
    // — otherwise a profile with that setting on would enter silent every
    // time until the toggle was flipped again in Settings.
    this.app.loadProfile(name);
    import("./mainMenu.js").then((m) => this.app.show(m.MainMenu));
  }

  _create() {
    const name = this.input.value.trim();
    if (!name) {
      this.error.textContent = "Please enter a name.";
      return;
    }
    if (storage.listProfiles().includes(name)) {
      this.error.textContent = "That profile already exists.";
      return;
    }
    storage.createProfile(name, this.pinInput.value);
    // New profiles only — existing profiles always skip straight to
    // MainMenu via _enter()/_choose() below, regardless of their
    // `onboarded` value, since profiles saved before this feature existed
    // have no `onboarded` field and would otherwise default to false.
    this.app.loadProfile(name);
    import("./onboarding.js").then((m) => this.app.show(m.Onboarding));
  }

  destroy() {}
}
